### PR TITLE
doc/dev/radosgw: start describing the bucket index

### PR DESCRIPTION
**Pull request by cbodley**
start documenting the rados bucket index and how the index transactions support our read-after-write consistency guarantee

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/dev/radosgw/bucket_index.rst
Comment (mattbenjamin) [inline doc/dev/radosgw/bucket_index.rst:46]: (recommend mentioning that this repair step is the so-called "dir-suggest" mechanism, to help developers)

**Review by mattbenjamin — COMMENTED** [review]
wow, great doc

**Review by ivancich — APPROVED** [review]
This is well-written and a huge help. The lack of such a document has been gaping hole in the developer docs for eons. Thanks, @cbodley!

File: doc/dev/radosgw/bucket_index.rst
Comment (cbodley) [reply doc/dev/radosgw/bucket_index.rst:5]: yeah, i wasn't quite sure where to put that. i think the `Sharding and Resharding` section and its discussion the bucket layout would be a good place to mention indexless as an 'index type'

Comment (ivancich) [reply doc/dev/radosgw/bucket_index.rst:5]: That seems like a good place. I think it still might make sense to at least add a note, as in "(note: there are also *indexless* buckets, see _SECTION-REF_ below)".

**Comment by cbodley** [comment]
filled in the `Sharding and Resharding` and `S3 Object Versioning` sections

**Comment by cbodley** [comment]
this still has an approval from @ivancich. would you or @mattbenjamin like to review further before merge?

**Review by anthonyeleven — APPROVED** [review]

**Pull request merged**
